libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+template <typename _Group>
+inline __device__
+void __cp_async_bulk_shared_global(const _Group &__g, char * __dest, const char * __src, size_t __size, uint64_t *__bar_handle) {
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk
+    if (__g.thread_rank() == 0) {
+        asm volatile(
+            "cp.async.bulk.shared::cluster.global.mbarrier::complete_tx::bytes [%0], [%1], %2, [%3];\n"
+            :
+            : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+              "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+              "r"(static_cast<_CUDA_VSTD::uint32_t>(__size)),
+              "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__bar_handle)))
+            : "memory");
     }
+}
+#endif // __CUDA_MINIMUM_ARCH__
 
-    template<thread_scope _Sco>
-    _LIBCUDACXX_INLINE_VISIBILITY static inline void __arrive_on(pipeline<_Sco> & __pipeline, async_contract_fulfillment __is_async) {
-        // pipeline does not sync on memcpy_async, defeat pipeline purpose otherwise
-        __unused(__pipeline);
-        __unused(__is_async);
+#if (defined(__CUDA_MINIMUM_ARCH__) && 800 <= __CUDA_MINIMUM_ARCH__) || (!defined(__CUDA_MINIMUM_ARCH__))
+template <size_t _Copy_size>
+inline __device__
+void __cp_async_shared_global(char * __dest, const char * __src) {
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async
+    static_assert(_Copy_size == 4 || _Copy_size == 8 || _Copy_size == 16, "cp.async.shared.global requires a copy size of 4, 8, or 16.");
+    asm volatile(
+        "cp.async.ca.shared.global [%0], [%1], %2, %2;"
+        :
+        : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+          "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+          "n"(_Copy_size)
+        : "memory");
+}
+
+template <>
+inline __device__
+void __cp_async_shared_global<16>(char * __dest, const char * __src) {
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async
+    // When copying 16 bytes, it is possible to skip L1 cache (.cg).
+    asm volatile(
+        "cp.async.cg.shared.global [%0], [%1], %2, %2;"
+        :
+        : "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__dest))),
+          "l"(static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__src))),
+          "n"(16)
+        : "memory");
+}
+
+template <size_t _Alignment, typename _Group>
+inline __device__
+void __cp_async_shared_global_impl(_Group __g, char * __dest, const char * __src, _CUDA_VSTD::size_t __size) {
+    static_assert(4 <= _Alignment, "cp.async requires at least 4-byte alignment");
+
+    // Maximal copy size is 16.
+    constexpr int __copy_size = (16 < _Alignment) ? 16 : _Alignment;
```

Comment:
I find it harder to read if the constraint is on the left
```suggestion
    constexpr int __copy_size = (_Alignment > 16) ? 16 : _Alignment;
```